xla/service/gpu/gemm_rewriter.cc

```diff
@@ -1164,6 +1182,53 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
     return true;
   }
 
+  absl::Status F8ScaleD(HloInstruction* instr, HloInstruction* existing_gemm,
+                        HloInstruction* d_scale) {
+    if (!ShapeUtil::IsScalar(d_scale->shape())) {
+      return OkStatus();
+    }
+
+    TF_ASSIGN_OR_RETURN(auto gpu_backend_config,
+                        existing_gemm->backend_config<GpuBackendConfig>());
+    const GemmBackendConfig& gemm_backend_config =
+        gpu_backend_config.gemm_backend_config();
+
+    // When the result of an FP8 GEMM is scaled but not type converted to FP8,
+    // the scaling factor must be forwarded to the Custom Call as a_scale
+    // (chosen here) or b_scale. The scaling factor is fused here when no input
+    // scaling factors were fused during the creation of the Custom Call. When
+    // the maximum of the absolute value of the result of an FP8 GEMM is
+    // calculated and the result is scaled and type converted to FP8, the
+    // scaling of the result is fused in F8ConvertD.
+    const int a_scale_operand_idx = gemm_backend_config.beta() == 0.0 ? 2 : 3;
+    if (!existing_gemm->operand(a_scale_operand_idx)->IsConstant() ||
+        existing_gemm->operand(a_scale_operand_idx)
+                ->literal()
+                .GetAsDouble({}) != 1.) {
+      return OkStatus();
+    }
+
+    // If necessary, invert the scaling factor of D and convert to F32.
+    if (instr->opcode() == HloOpcode::kDivide) {
+      Literal one_literal = LiteralUtil::One(d_scale->shape().element_type());
+      HloInstruction* one = existing_gemm->parent()->AddInstruction(
+          HloInstruction::CreateConstant(one_literal.Clone()));
+      d_scale = existing_gemm->AddInstruction(HloInstruction::CreateBinary(
+          d_scale->shape(), HloOpcode::kDivide, one, d_scale));
```

Comment:
When creating HLO instructions, please copy the metadata from the original instruction. See line 761 for an example.
Same below.